alacritty_terminal/src/term/mod.rs

```diff
@@ -941,6 +964,105 @@ impl Term {
         self.dirty
     }
 
+    pub fn get_damage(&mut self) -> Vec<DamageRect> {
+        let very_dirty = self.grid.very_dirty;
+        let cur_cursor = self.cursor;
+        let last_cursor = self.last_cursor;
+        self.last_cursor = cur_cursor;
+        let cursor_moved = last_cursor.point != cur_cursor.point;
+        self.grid.very_dirty = false;
+
+        if very_dirty {
+            // The grid is very dirty, probably through scrolling.
+            // Clear all dirty bits and return a full damage rect.
+
+            for line in 0..self.grid.num_lines().0 {
+                for col in 0..self.grid.num_cols().0 {
+                    self.grid[Line(line)][Column(col)].clear_dirty();
+                }
+            }
+            return vec![
+                DamageRect{
+                    x: 0,
+                    y: 0,
+                    width: self.grid.num_cols().0,
+                    height: self.grid.num_lines().0,
+                }
+            ];
+        }
+
+        // We need to damage the selection, so compute it.
+        // TODO: Only track when selection changes.
+        let selection_range = self.grid.selection.as_ref().and_then(|s| s.to_span(self)).map(|span| {
+            let (limit_start, limit_end) = if span.is_block {
+                (span.end.col, span.start.col)
+            } else {
+                (Column(0), self.cols() - 1)
+            };
+
+            // Get on-screen lines of the selection's locations
+            let mut start = self.buffer_to_visible(span.start);
+            let mut end = self.buffer_to_visible(span.end);
+
+            // Start and end lines are swapped as we switch from buffer to line coordinates
+            if start > end {
+                mem::swap(&mut start, &mut end);
+            }
+
+            // Trim start/end with partially visible block selection
+            start.col = max(limit_start, start.col);
+            end.col = min(limit_end, end.col);
+
+            SelectionRange::new(start.into(), end.into(), span.is_block)
+        });
+
+        // Gather damage
+        let mut damage = Vec::new();
+        for line in 0..self.grid.num_lines().0 {
+            let mut dirty = false;
+            let mut dirty_col = 0;
+            for col in 0..self.grid.num_cols().0 {
+                let is_selected = selection_range
+                    .as_ref()
+                    .map(|range| range.contains(index::Column(col), index::Line(line)))
+                    .unwrap_or(false);
+
+                let index = Linear::new(self.grid.num_cols(), index::Column(col), index::Line(line));
+                let is_highlighted = self.grid.url_highlight.as_ref().map(|range| range.contains_(index)).unwrap_or(false);
+                let is_old_cursor = last_cursor.point.line.0 == line && last_cursor.point.col.0 == col;
```

Comment:
I recommend you to always redraw cursor, since we'll want to redraw it most of the time, but the logic should become much simpler to handle. @chrisduerr , what do you think about it?

Reply:
Mouse motion over URLs and selection would be some relevant occasions that come to my mind where the cursor will not change.

Reply:
Yeah, but it's rather and exception here, since for now cursor handling is kind of a major problem on this PR. Since it can be dynamically changed and so on, so probably updating it all every time worth it.

Reply:
Cursor, selection and url_highlight are weak spots, as they are all effectively cell state tracked out-of-band.